src/checkmatelib/url/canonicalize.py

```diff
@@ -52,7 +66,12 @@ def canonical_split(cls, url):
         # 127, #, or %. The escapes should use uppercase hex characters.
         netloc = cls._partial_quote(netloc)
         path = cls._partial_quote(path)
+
+        # Make a distinction between an empty query and no query at all. This
+        # relies on us not having a fragment
         query = cls._partial_quote(query)
+        if not query:
+            query = "" if url.endswith("?") else None
```

Comment:
As a nice bonus, if you do just whack this into `urllib.unparse` it doesn't care about this distinction, so we can maintain compatibility with this.